src/Processors/Transforms/PlanSquashingTransform.cpp

```diff
@@ -0,0 +1,180 @@
+{
+}
+
+IProcessor::Status PlanSquashingTransform::prepare()
+{
+    Status status = Status::Ready;
+
+    while (planning_status != PlanningStatus::FINISH)
+    {
+        switch (planning_status)
+        {
+            case INIT:
+                init();
+                break;
+            case READ_IF_CAN:
+                status = prepareConsume();
+                break;
+            case WAIT_IN:
+                planning_status = PlanningStatus::READ_IF_CAN;
+                return Status::NeedData;
+            case PUSH:
+                return sendOrFlush();
+            case FLUSH:
+                return sendOrFlush();
+            case FINISH:
+                break; /// never reached
+        }
+    }
+    if (status == Status::Ready)
+        status = finish();
+    else
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "There should be a Ready status to finish the PlanSquashing");
+
+    return status;
+}
+
+void PlanSquashingTransform::work()
+{
+    prepare();
+}
+
+void PlanSquashingTransform::init()
+{
+    for (auto input: inputs)
+        if (!input.isFinished())
+            input.setNeeded();
+
+    planning_status = PlanningStatus::READ_IF_CAN;
+}
+
+IProcessor::Status PlanSquashingTransform::prepareConsume()
+{
+    bool inputs_have_no_data = true, all_finished = true;
+    for (auto & input : inputs)
+    {
+        if (!input.isFinished())
+            all_finished = false;
+
+        if (input.hasData())
+        {
+            inputs_have_no_data = false;
+            chunk = input.pull();
+            transform(chunk);
+
+            if (chunk.hasChunkInfo())
+            {
+                planning_status = PlanningStatus::PUSH;
+                return Status::Ready;
+            }
+        }
+    }
+
+    if (all_finished) /// If all inputs are closed, we check if we have data in balancing
+    {
+        if (balance.isDataLeft()) /// If we have data in balancing, we process this data
+        {
+            planning_status = PlanningStatus::FLUSH;
+            flushChunk();
+            return Status::Ready;
+        }
+        planning_status = PlanningStatus::PUSH;
+        return Status::Ready;
+    }
+
+    if (inputs_have_no_data)
+        planning_status = PlanningStatus::WAIT_IN;
+
+    return Status::Ready;
+}
+
+IProcessor::Status PlanSquashingTransform::waitForDataIn()
+{
+    bool all_finished = true;
+    bool inputs_have_no_data = true;
+    for (auto & input : inputs)
+    {
+        if (input.isFinished())
+            continue;
+
+        all_finished = false;
+
+        if (input.hasData())
+            inputs_have_no_data = false;
+
+    }
+    if (all_finished)
+    {
+        planning_status = PlanningStatus::READ_IF_CAN;
+        return Status::Ready;
+    }
+
+    if (!inputs_have_no_data)
+    {
+        planning_status = PlanningStatus::READ_IF_CAN;
+        return Status::Ready;
+    }
+
+    return Status::NeedData;
+}
+
+void PlanSquashingTransform::transform(Chunk & chunk_)
```

Comment:
lets do it a way

`Chunk PlanSquashingTransform::transform(Chunk && chunk_)`
`Chunk PlanSquashingTransform::flushChunk()`